File: 3_longest_substring_without_repeating_characters/main.py
# -*- coding: utf-8 -*-


class Solution(object):

    def length_of_longest_substring(self, s):
        i, j = 0, 1
        max_len = 0
        size = len(s)
        shown = dict()

        while i < size:
            shown[s[i]] = True
            while j < size:
                if shown.get(s[j], False):
                    break
                shown[s[j]] = True
                j += 1

            max_len = max(j - i, max_len)
            if j >= size:
                break
            shown[s[i]] = False
            i += 1
            if i >= j:
                j = i + 1

        return max_len


if __name__ == '__main__':
    sl = Solution()

    s = 'abcabcbb'
    print(sl.length_of_longest_substring(s))

    s = 'bbbbb'
    print(sl.length_of_longest_substring(s))

    s = 'pwwkew'
    print(sl.length_of_longest_substring(s))

File: 3_longest_substring_without_repeating_characters/go/main.go
package main

import (
	"fmt"
)

func lengthOfLongestSubstring(s string) int {
	var maxLen int

	sRunes := []rune(s)
	i, j, size := 0, 1, len(sRunes)
	shown := make(map[rune]bool)

	for i < size {
		shown[sRunes[i]] = true
		for j < size {
			if shown[sRunes[j]] {
				break
			}
			shown[sRunes[j]] = true
			j++
		}
		if (j - i > maxLen) {
			maxLen = j - i
		}
		if (j >= size) {
			break
		}
		shown[sRunes[i]] = false
		i++
		if i >= j {
			j = i + 1
		}
	}

	return maxLen
}

func main() {
	s := "abcabcbb"
	fmt.Println(lengthOfLongestSubstring(s))

	s = "bbbbb"
	fmt.Println(lengthOfLongestSubstring(s))

	s = "pwwkew"
	fmt.Println(lengthOfLongestSubstring(s))
}
